# Epigenetic-Clock
This program analyzes methylation levels at six CpG sites in the genome to produce a prediction of an individual's biological age, using different machine learning and deep learning models.
## Background
Scientists define there to be two main types of age: chronological, and biological age. Chronological age is based on how long you have been alive (years since birth), whereas biological age is a rough estimate of how healthy your body is by measuring various different biomarkers. For more information on this, check out my recent article on biological age right here: [What is Biological Age?](https://medium.com/biotein/what-is-biological-age-383692f53e60)

One way of determining biological age (as also described in the article) is to measure the methylation levels in your genome. At sites known as CpG sites, along your DNA, various proteins can add on or remove methyl groups, in order to control which genes are expressed into proteins, and which ones aren't. As you age, however, the systems responsible for maintaining this control of the genome begin to break down, leading to errors in methylation. As such, some sites begin to see an accumulation of methyl tags, while others have their tags removed, as you age.

Dr. Steve Horvath, longevity researcher at University of California Los Angeles, used this progression to develop a system known as the Horvath aging clock, which is able to produce an accurate estimate of your biological age. For example, after analyzing the methylation levels in a sample of your cells, the clock might tell you that your biological age is around 30, which basically means that your methylation levels (and your body's health) is similar to the average 30-year-old. If your actual age is somewhere around 40, this means you are living a healthy life, whereas if you are only 20 years old, then this suggests that you are living an unhealthy life. This information thus equips people with the knowledge to take control of their lifestyle and live a healthy life.

## Brief Explanation of This Project
In this project, I aim to somewhat replicate the epigenetic clock developed by Horvath, using methylation data taken from blood samples to predict an individual's age. The datasets and methylation sites chosen for this project were advised by the research paper titled [Human Age Prediction Based on DNA Methylation Using a Gradient Boosting Regressor](https://doi.org/10.3390/genes9090424). There are two notable differences between the paper and this code:
1. Since this code also takes into account the gender of the patients whose methylation data is available, GSE30870, which does not contain gender information in the dataset, was removed from the sample.
2. This notebook contains three different ML Regression algorithms and a Deep Neural Network, and compares the performance of each model on the training and testing dataset. The four models developed in this notebook are: Multivariable Linear Regression, Random Forest Regression, Gradient Boosting Regression, and a Deep Neural Network.

## Data Summary
The data was accessed from: [GSE20236](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20236), [GSE20242](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20242), [GSE20797](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE27097), [GSE27317](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE27317), [GSE32149](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE32149), [GSE34257](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34257), [GSE34869](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE34869), [GSE36064](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36064), [GSE36642](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36642), [GSE36812](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36812), [GSE37008](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37008), [GSE40279](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40279), [GSE41169](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE41169), [GSE53128](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE53128), [GSE65638](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65638), the healthy datasets listed in the paper, minus GSE30870.

When accessing these datasets from their sources using R (code not included in this repository), I made sure to only store the columns for the six CpG sites listed in the paper, which were found to be correlated with ageing in the paper using Pearson correlation analysis. The sites are as listed: cg09809672, cg22736354, cg02228185, cg01820374, cg06493994, cg19761273.
The original GSE datasets also had gender as a categorical variable (Male or Female), but, in order to carry out regression on the data, this must be converted into numerical form. As thus, in the gender column in the set, 1 represents Females, whereas 0 represents Males.
Other forms of data cleaning, like removing text from the age column, were also required, but are not crucial enough to delve into.

The folder Subsets contains all the datasets used to form the overall final dataset, Methylation_Dataset.csv. The final dataset has 1440 patients, but many of these patients lack some crucial data (such as methylation levels and age), and are thus are dropped in the code, bringing the patient count down to 1334 patients.
